### PR TITLE
Add mermaids for file content encryption.

### PR DESCRIPTION
**Pull request by chenkins**

### CODE REVIEW

File: file content encryption/AES-256-GCM.md [hunk 1]
```diff
@@ -30,12 +51,43 @@ cleartextBlocks[] := split(data: cleartext, maxBytes: n)
...
+```mermaid
+---
+title: File Content Encryption for AES-256-GCM-XXk format
+---
+erDiagram
+    ENCRYPTEDFILECONTENT ||--|| GENERALHEADERFIELDS: has
+    ENCRYPTEDFILECONTENT ||--|| CUSTOMHEADERFIELDS: has
+    ENCRYPTEDFILECONTENT ||--|{ CIPHERTEXTBLOCK: "has"
+    ENCRYPTEDFILECONTENT["encrypted file content"]
+
```
Comment (overheadhunter) [inline file content encryption/AES-256-GCM.md]: I think another intermediate step to separate header from body will make it even easier to understand:

```suggestion
    FILE["encrypted file"]
    FILE_HEADER["file header"]
    FILE_BODY["encrypted body"]
    FILE ||--|| FILE_HEADER: has
    FILE ||--|| FILE_BODY: has

    FILE_HEADER ||--|| GENERALHEADERFIELDS: has
    FILE_HEADER ||--|| CUSTOMHEADERFIELDS: has

    FILE_BODY ||--|{ CIPHERTEXTBLOCK: "consists of"

```

File: file name encryption/AES-SIV-512-B64URL.md [hunk 2]
```diff
@@ -79,6 +79,66 @@ let dirPath = vaultRoot + '/d/' + dirIdString[0..2] + '/' + dirIdString[2..32]
...
+   dirId --> concat
+   concat{{concat}}
+   concat -->|cleartextBlocks:| fileContentEncryption
+   fileContentEncryption{{file content encryption}}
+   directorySeed -->|seed:| fileContentEncryption
+   fileContentEncryption --> dirUvf
+   dirUvf["directory metadata dir.uvf content"]
+```
+
+
```
Comment (overheadhunter) [inline file name encryption/AES-SIV-512-B64URL.md]: I like the idea, but let's break this into two charts to make it more digestible:

1. decide which dirId and seed is used
2. derive dir path and encrypt node names

File: file name encryption/AES-SIV-512-B64URL.md [hunk 3]
```diff
@@ -79,6 +79,66 @@ let dirPath = vaultRoot + '/d/' + dirIdString[0..2] + '/' + dirIdString[2..32]
...
+   aesSiv --> base64Url
+   base64Url{{base64url}}
+   base64Url --> join
+   uvf --> join
+   uvf["'.uvf'"]
+   join{{"join"}}
+   join --> ciphertextName
+   directorySeed --> concat
+   dirId --> concat
+   concat{{concat}}
```
Comment (overheadhunter) [inline file name encryption/AES-SIV-512-B64URL.md]: This is incorrect. We do not concatenate seed and dirId. Instead we create a normal encrypted file whose sole content is the `dirId`. The seed is used in the file header (as correctly indicated by `directorySeed -->|seed:| fileContentEncryption`)

**Review by overheadhunter — CHANGES_REQUESTED** [review]
Thanks for taking the initiative, these diagrams are really helpful!

I noticed you renamed `fileKey` → `fileContentKey` as well as KDF's param `secret` → `seed`. While I'm not completely opposed to renaming, I think this is an unrelated change and I would at least like to know the motivation behind this. Note that in the context of KDFs the secret param is often referred to as `input key material` or just `input` but I didn't stumble upon `seed` yet. To make reviews easier, I tend to stick with standard names.

File: file content encryption/AES-256-GCM.md (hunk 1 above)
Comment (chenkins) [reply file content encryption/AES-256-GCM.md]: Relates to [comment in #24](https://github.com/encryption-alliance/unified-vault-format/pull/24#discussion_r1719526575)

File: file name encryption/AES-SIV-512-B64URL.md (hunk 3 above)
Comment (chenkins) [reply file name encryption/AES-SIV-512-B64URL.md]: I was mislead by

> Every directory requires certain metadata that affects the file name encryption of its direct children:
> 
> * the seed used to derive keys
> * the directory ID

into thinking that both go in to the body. However, it's actually clear later on:

> The cleartext content of `dir.uvf` is the 32 byte dirId.

Comment (chenkins) [reply file name encryption/AES-SIV-512-B64URL.md]: Done: 5ca480306b88a46cf711b845e89a2d5bdd2dcdb9

File: file name encryption/AES-SIV-512-B64URL.md (hunk 2 above)
Comment (chenkins) [reply file name encryption/AES-SIV-512-B64URL.md]: Done: 92dbd884e51ba0dac60d8b024af0f6ff702a6bed

File: file content encryption/AES-256-GCM.md [hunk 4]
```diff
@@ -13,8 +14,28 @@ The header needs to be encrypted using a 256 bit key derived from the seed using
 headerKey := kdf(secret: latestSeed, length: 32, context: "fileHeader")
 headerNonce := csprng(bytes: 12)
 fileKey := csprng(bytes: 32)
-encryptedFileKey, tag := aesGcm(cleartext: fileKey, key: headerKey, nonce: headerNonce, ad: generalHeaderFields)
-header := generalHeaderFields . headerNonce . encryptedFileKey . tag
+encryptedfileKey, tag := aesGcm(cleartext: fileKey, key: headerKey, nonce: headerNonce, ad: generalHeaderFields)
+header := generalHeaderFields . headerNonce . encryptedfileKey . tag
```
Comment (overheadhunter) [inline file content encryption/AES-256-GCM.md]: revert change (broken camel case)

File: file name encryption/AES-SIV-512-B64URL.md [hunk 5]
```diff
@@ -181,4 +181,69 @@ Thus, for a given cleartext directory structure like this...
...
+    csprng32 --> dirId
+    csprng32{{"csprng(32)"}}
+    initialSeed --> directorySeed
+    initialSeed -->|secret:| kdfRootDirId
+    kdfRootDirId{{"kdf(secret,32,'rootDirId')"}}
+    kdfRootDirId --> dirId
+    subgraph Outputs
+        directorySeed
+        dirId
+    end
```
Comment (overheadhunter) [inline file name encryption/AES-SIV-512-B64URL.md]: how about this?

```suggestion
---
title: Choosing a Seed and Directory ID
---
flowchart TD
    subgraph "Which Seed?"
        isRoot1{is root dir?}
        isRoot1 -->|y| initialSeed
        isRoot1 -->|n| isExistingDir1

        isExistingDir1{is existing dir?}
        isExistingDir1 -->|y| readSeed
        isExistingDir1 -->|n| latestSeed

        readSeed{{"read seed from `dir.uvf`"}}
        readSeed --> dirFileSeed[seed from dir.uvf]
    end

    subgraph "Which dirId?"
        isRoot2{is root dir?}
        isRoot2 -->|n| isExistingDir2
        isRoot2 -->|y| kdfRootDirId

        isExistingDir2{is existing dir?}
        isExistingDir2 -->|y| readDirId
        isExistingDir2 -->|n| csprng32
        
        csprng32 --> randomDirId[random dirId]
        csprng32{{"csprng(32)"}}

        initialSeed -->|secret:| kdfRootDirId    
        kdfRootDirId{{"kdf(secret,32,'rootDirId')"}}
        kdfRootDirId --> rootDirId[root dirId]

        readDirId{{"read dirId from `dir.uvf`"}}
        readDirId --> dirFileId[dirId from dir.uvf]
    end
```

File: file name encryption/AES-SIV-512-B64URL.md [hunk 6]
```diff
@@ -181,4 +181,69 @@ Thus, for a given cleartext directory structure like this...
...
+   aesSiv{{aesSiv}}
+   parentDirId -->|ad:| aesSiv
+   clearTextName -->|secret:| aesSiv
+   aesSiv --> base64Url
+   base64Url{{base64url}}
+   base64Url --> join
+   uvf --> join
+   uvf["'.uvf'"]
+   join{{"join"}}
+   join --> ciphertextName
```
Comment (overheadhunter) [inline file name encryption/AES-SIV-512-B64URL.md]: graph direction is odd (looks like `join('.uvf', base64)`). let's simplify:

```suggestion
   base64Url --> appendFileExt
   appendFileExt{{"append '.uvf'"}}
   appendFileExt --> ciphertextName
```

File: file name encryption/AES-SIV-512-B64URL.md [hunk 7]
```diff
@@ -181,4 +181,69 @@ Thus, for a given cleartext directory structure like this...
...
+   truncate --> base32
+   base32{{base32}}
+   base32 --> dirIdString
+   dirIdString --> head
+   head{{"_[0..2]"}}
+   dirIdString --> tail
+   tail{{"_[2..32]"}}
+   head -->|$0:| pattern
+   tail -->|$1:| pattern
+   pattern{{"'d/'$0'/'$1"}}
```
Comment (overheadhunter) [inline file name encryption/AES-SIV-512-B64URL.md]: let's assume a language that can deal with a simpler pattern string 😉

```suggestion
   pattern{{"d/$0/$1"}}
```

**Review by overheadhunter — CHANGES_REQUESTED** [review]

File: file content encryption/AES-256-GCM.md (hunk 4 above)
Comment (chenkins) [reply file content encryption/AES-256-GCM.md]: 91951506efac73eebcb0b8b0bed5508a53f138c8

**Review by overheadhunter — APPROVED** [review]

**Comment by overheadhunter** [comment]
feel free to merge when you're ready

**Pull request merged**